File: app/src/main/java/com/example/movieapi/adapter/MovieItemAdapter.java
package com.example.movieapi.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;

import com.example.movieapi.R;
import com.example.movieapi.model.MovieItem;
import com.squareup.picasso.Picasso;

public class MovieItemAdapter extends ListAdapter<MovieItem, MovieItemViewHolder> {

    View.OnClickListener onClickListener;

    public MovieItemAdapter(@NonNull DiffUtil.ItemCallback<MovieItem> diffCallback, View.OnClickListener onClickListener) {
        super(diffCallback);
        this.onClickListener = onClickListener;
    }

    @NonNull
    @Override
    public MovieItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.movie_item, parent, false);

        return new MovieItemViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull MovieItemViewHolder holder, int position) {

        MovieItem movieItem = getItem(position);

        holder.titleTextView.setText(movieItem.getTitle());
        holder.yearTextView.setText(movieItem.getYear());
        holder.ratingTextView.setText(movieItem.getRating());

        String posterUrl = "https://image.tmdb.org/t/p/w500" + movieItem.getPosterUrl();
        Picasso.get().load(posterUrl).into(holder.posterImageView);

        holder.itemCardView.setOnClickListener(onClickListener);

    }
}

File: app/src/main/java/com/example/movieapi/adapter/MovieItemAdapter.kt
package com.example.movieapi.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.example.movieapi.R
import com.example.movieapi.data.OnItemClick
import com.example.movieapi.model.MovieItem
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.movie_item.view.*

class MovieItemAdapter(
        diffCallback: DiffUtil.ItemCallback<MovieItem>,
        private val onClickListener: OnItemClick
) : ListAdapter<MovieItem, MovieItemViewHolder>(diffCallback) {

    companion object {
        private const val BASE_IMAGE_URL = "https://image.tmdb.org/t/p/w500"
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieItemViewHolder =
            MovieItemViewHolder(
                    LayoutInflater.from(parent.context)
                            .inflate(R.layout.movie_item, parent, false))

    override fun onBindViewHolder(holder: MovieItemViewHolder, position: Int) {
        with(holder.itemView) {
            getItem(position)?.let { item ->
                titleTextView.text = item.title
                yearTextView.text = item.year
                ratingTextView.text = item.rating
                val posterUrl: String = BASE_IMAGE_URL + item.posterUrl
                Picasso.get().load(posterUrl).into(posterImageView)
                itemCardView.setOnClickListener { onClickListener.onClick(item.id) }
            }
        }
    }

}